soda/bigquery/soda/data_sources/bigquery_data_source.py

```diff
@@ -94,9 +99,24 @@ def __init__(self, logs: Logs, data_source_name: str, data_source_properties: di
                 scopes=self.auth_scopes,
             )
 
+        if self.data_source_properties.get("use_context_auth"):
+            self.logs.info("Using application default credentials.")
+            self.credentials, _ = default()
+
+        impersonation_account = self.data_source_properties.get("impersonation_account")
+        if impersonation_account:
+            self.logs.info("Using impersonation of Service Account.")
+            self.credentials = impersonated_credentials.Credentials(
+                source_credentials=self.credentials,
+                target_principal=str(impersonation_account),
+                target_scopes=self.data_source_properties.get("scopes", default_auth_scopes),
+            )
+
         # All remaining parameters
-        # Usually the project_id comes from the self.account_info_dict
-        self.project_id = self.account_info_dict.get("project_id") if self.account_info_dict else None
+        # Usually the project_id comes from the self.account_info_dict or the credentials itself
+        self.project_id = (
+            self.account_info_dict.get("project_id") if self.account_info_dict else self.credentials.quota_project_id
```

Comment:
I dont think `self.credentials.quota_project_id` is safe to do without checking if its there, if default credentials is used then project id might be `None`:
(from google lib doc):
```
the current environment's credentials and project ID. Project ID
may be None, which indicates that the Project ID could not be
ascertained from the environment.
```

Reply:
From how I read the Docs, the `self.credentials.quota_project_id` obtained from the [Credentials Module](https://google-auth.readthedocs.io/en/master/reference/google.auth.credentials.html#google.auth.credentials.Credentials.quota_project_id) is always set.   

Its however different from the `project_id` obtained from `google.auth.default`. As you said, this 
```
Returns the current environment's credentials and project ID. Project ID
may be None, which indicates that the Project ID could not be
ascertained from the environment.
```

By default and when using context auth, the quota_project_id parameter will be used.
The user however can optionally overwrite in the connection properties.
See:
`self.project_id = data_source_properties.get("project_id", self.project_id)`

Reply:
@m1n0 Following suggestions:

1. We could try receive the `project_id` from the `google.auth.default()`
2. If its not there, fall back to using `quota_project_id`
Optionally: User can set `project_id` via connection properties

Reply:
I see, that makes sense, lets make sure that the option to set the project_id via connection properties is there as well!